Updates to server documentation (#210)
temporalio/documentation: docs/server-options.md

---
id: server-options
title: Temporal server options
sidebar_label: Start options
---

## Overview

> To run the Temporal, include the temporal server package `go.temporal.io/server/temporal` in your Go application. Then use it to create and start a new server:

Reviewer: we need to start positioning Temporal to be usable by people who don't use Go eventually. suggest this wording

```suggestion
To run Temporal as a production Go application, include the temporal server package `go.temporal.io/server/temporal`. Then use it to create and start a new server:
```

```go
s := temporal.NewServer()
err := s.Start()
if err != nil{
	log.Fatal(err)
}
```

> ## Options

Reviewer: lets try to guide the user as much as possible on what most users may need vs what a select few may need. may want to split out "options" and "advanced options"?

Author: I like idea - not sure exactly what "advanced" would mean in this context? Any one of these could range from simple to complex in terms of usage and consequences. I think right now, lets just make it available as a reference.

Reviewer: @flossypurse  ok it was just an idea. i like being guided to "options all of you definitely want to know about" vs "options most of you can ignore"

Author: Added a task to revisit this - I think it will depend on the additional context we get for each of the options.

`NewServer()` accepts functions as parameters. Each function returns a `ServerOption` that is applied to the instance.

### WithConfig

Overrides the default configuration with a custom configuration that is defined in the config package `go.temporal.io/server/common/service/config`.

```go
s := tempora.NewServer(
	temporal.WithConfig(cfg),
)
```

### WithConfigLoader

Load a custom configuration from a file.

```go
s := tempora.NewServer(

> 	temporal.WithConfigLoader(configDir, env, zone),

Reviewer: maybe spell out where `configDir` is

Author: Eventually, these will be converted to real samples via snipsync which will include a link to the source which will create a traceable path to these parameters.
For now, I have added a link to the server source code with the functions.

)
```

### ForServices

Sets the list of all valid temporal services. The default can be used from the `go.temporal.io/server/temporal` package.

```go
s := tempora.NewServer(
	temporal.ForServices(temporal.Services),
)

> ```

Reviewer: suggest providing other common usages of this, showing default alone doesnt seem helpful?

Author: Tracking as a task

### InterruptOn

> This option provides a channel that interrupts the server on the signal from that channel.

Reviewer: maybe spell out usecase - when is this helpful? is this a rare usecase?

Author: For these server options, I will need to get more input from the team.
In case I don't get this info in a timely manner, I created a task in our board to track.

Reviewer: wonderful

- If `temporal.InterruptOn()` is not passed, `server.Start()` is not blocking and `server.Stop()` needs to be called somewhere.
- If `temporal.InterruptOn(nil)` is passed, `server.Start()` blocks forever until the process is killed.
- If `temporal.InterruptOn(temporal.InterruptCh())` is passed, `server.Start()` blocks until Ctrl+C is used and then the server is gracefully shut down.
- If `temporal.Interrupt(someCustomChan)` is passed, `server.Start()` blocks until a signal is sent to `someCustomChan`.

```go
s := temporal.NewServer(
	temporal.InterruptOn(temporal.InterruptCh()),
)
```

### WithAuthorizer

Sets a low level authorizer that determines whether to allow or deny inbound API calls.

```go
s := temporal.NewServer(

> 	temporal.WithAuthorizer(myAuthorizer),

Reviewer: link to docs on writing authorizers?

)
```

### WithTLSConfigFactory

Overrides the default TLS configuration provider. `TLSConfigProvider` is defined in the `go.temporal.io/server/common/rpc/encryption` package.

```go
s := temporal.NewServer(
	temporal.WithTLSConfigFactory(yourTLSConfigProvider),
)
```

### WithClaimMapper

Configures a role mapper for authorization.

```go
s := temporal.NewServer(
  temporal.WithClaimMapper(func(cfg *config.Config) authorization.ClaimMapper {
		return authorization.NewDefaultJWTClaimMapper(
			authorization.NewRSAKeyProvider(cfg),
			cfg
		)
	}),
)
```

### WithCustomMetricsReporter

Sets a custom tally metric reporter.

```go
s := temporal.NewServer(
	temporal.WithCustomMetricsReporter(myReporter),
)

> ```

Reviewer: what reporters do we recommend?

Author: Tracking as a task.

---
id: server-options
title: Temporal server options
sidebar_label: Start options
---

> ## Overview

Reviewer 2: I think we should mention that by default you don't need any of these and the default docker build should be used. 

Only if you want to compile the service yourself to override certain components this approach should be used. And we want to list up front in which specific scenarios it can be useful.
